### PR TITLE
Update multi-valued claims separately in SCIM user

### PR DESCRIPTION
**Pull request by AnuradhaSK**
### Purpose
Fix for wso2/product-is#8135
Public fix of https://github.com/wso2-support/identity-inbound-provisioning-scim2/pull/94

### When to merge
Depends on https://github.com/wso2/carbon-kernel/pull/2705/
https://github.com/wso2/charon/pull/286

### CODE REVIEW

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java [hunk 1]
```diff
@@ -3672,6 +3833,116 @@ private void updateUserClaims(User user, Map<String, String> oldClaimList,
...
+     * @param claimMap Map of claim URIs against claim value as string.
+     * @return Map of claim URIs against claim value as a list of string.
+     */
+    private Map<String, List<String>> convertClaimValuesToList(Map<String, String> claimMap) {
+
+        Map<String, List<String>> claimMapWithListValues = new HashMap<>();
+        String claimValueSeparator = ",";
+        if (StringUtils.isNotEmpty(FrameworkUtils.getMultiAttributeSeparator())) {
+            claimValueSeparator = FrameworkUtils.getMultiAttributeSeparator();
+        }
```
Comment (madurangasiriwardena) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java:3938]:
```suggestion
        String claimValueSeparator = FrameworkUtils.getMultiAttributeSeparator();
        if (StringUtils.isEmpty(claimValueSeparator)) {
            claimValueSeparator = ",";
        }
```

Comment (AnuradhaSK) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java]: Fixed in 0b1de97

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java [hunk 2]
```diff
@@ -877,6 +878,166 @@ public User updateUser(User user, Map<String, Boolean> requiredAttributes) throw
...
+             /* Set thread local property to signal the downstream SCIMUserOperationListener
+                about the provisioning route. */
+            SCIMCommonUtils.setThreadLocalIsManagedThroughSCIMEP(true);
+            // Get user claim values.
+            Map<String, String> claims = AttributeMapper.getClaimsMap(user);
+            // Get claims mapped with simple multi-valued attributes.
+            Map<String, String> allSimpleMultiValuedClaims = new HashMap<>();
+
+            for (String simpleMultiValuedAttribute : allSimpleMultiValuedAttributes) {
+                allSimpleMultiValuedClaims.put(simpleMultiValuedAttribute, "");
```
Comment (somindatommy) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java:908]: Shall we use String.EMPTY here?

Comment (somindatommy) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java:908]: Shall we use String.EMPTY here?

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java [hunk 3]
```diff
@@ -877,6 +878,166 @@ public User updateUser(User user, Map<String, Boolean> requiredAttributes) throw
...
+            // Check if username of the updating user existing in the userstore.
+            try {
+                String userStoreDomainFromSP = getUserStoreDomainFromSP();
+                SCIMResourceTypeSchema schema = SCIMResourceSchemaManager.getInstance().getUserResourceSchema();
+                User oldUser = this.getUser(user.getId(), ResourceManagerUtil.getAllAttributeURIs(schema));
+                if (userStoreDomainFromSP != null && !userStoreDomainFromSP
+                        .equalsIgnoreCase(IdentityUtil.extractDomainFromName(oldUser.getUserName()))) {
+                    String errorMessage =
+                            String.format("User :  %s is not belong to userstore %s. Hence user updating fail",
+                                    oldUser.getUserName(), userStoreDomainFromSP);
```
Comment (somindatommy) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java:922]: `User :  %s  does not belong to userstore: %s. Hence user updating failed`

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java [hunk 4]
```diff
@@ -877,6 +878,166 @@ public User updateUser(User user, Map<String, Boolean> requiredAttributes) throw
...
+                    throw new CharonException(errorMessage);
+                }
+                if (getUserStoreDomainFromSP() != null &&
```
Comment (somindatommy) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java:925]: We already have a variable with the value from `getUserStoreDomainFromSP()` method. Can we use that?
I think we can improve the logic using nested If conditions.

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java [hunk 5]
```diff
@@ -877,6 +878,166 @@ public User updateUser(User user, Map<String, Boolean> requiredAttributes) throw
...
+                        !UserCoreConstants.PRIMARY_DEFAULT_DOMAIN_NAME.equalsIgnoreCase(getUserStoreDomainFromSP())) {
+                    user.setUserName(IdentityUtil
+                            .addDomainToName(UserCoreUtil.removeDomainFromName(user.getUserName()),
+                                    getUserStoreDomainFromSP()));
+                }
+                // This is handled here as the IS is still not capable of updating the username via SCIM.
+                if (!StringUtils.equals(user.getUserName(), oldUser.getUserName())) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Failing the request as attempting to modify username. Old username: "
```
Comment (somindatommy) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java:934]: `BAD REQUEST. User: <old username> attempting to modify the username as : <new username>` WDYT?

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java [hunk 6]
```diff
@@ -877,6 +878,166 @@ public User updateUser(User user, Map<String, Boolean> requiredAttributes) throw
...
+                                + oldUser.getUserName() + ", new username: " + user.getUserName());
+                    }
+
```
Comment (somindatommy) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java:937]: Shall we remove extra new lines

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java [hunk 7]
```diff
@@ -877,6 +878,166 @@ public User updateUser(User user, Map<String, Boolean> requiredAttributes) throw
...
+            boolean isExistingUser;
+            if (StringUtils.isNotEmpty(user.getId())) {
+                isExistingUser = carbonUM.isExistingUserWithID(user.getId());
+            } else {
+                isExistingUser = carbonUM.isExistingUser(user.getUserName());
+            }
+
+            if (!isExistingUser) {
+                throw new CharonException("User name is immutable in carbon user store.");
+            }
```
Comment (somindatommy) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java:954]: Can we move this to a new method? Such as `validateUserExistance` WDYT?

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java [hunk 8]
```diff
@@ -877,6 +878,166 @@ public User updateUser(User user, Map<String, Boolean> requiredAttributes) throw
...
+
+            updateUserClaims(user, oldClaimList, claimValuesInLocalDialect, allSimpleMultiValuedClaimsList);
+
+            if (log.isDebugEnabled()) {
+                log.debug("User: " + user.getUserName() + " updated through SCIM.");
+            }
+            return getUser(user.getId(), requiredAttributes);
+        } catch (UserStoreException e) {
+            log.error("Error while updating attributes of user: " + user.getUserName(), e);
```
Comment (somindatommy) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java:1021]: Avoid logging the error and throwing it the same time.

Comment (somindatommy) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java:1021]: Shall we fix it in other places?

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java [hunk 9]
```diff
@@ -3672,6 +3833,116 @@ private void updateUserClaims(User user, Map<String, String> oldClaimList,
...
+                            (List<String>) CollectionUtils.subtract(newValue, oldValue));
+                }
+            } else if (oldClaimList.containsKey(key) && !newClaimList.containsKey(key)) {
+                simpleMultiValuedClaimsToBeRemoved.put(key, Arrays.asList(oldClaimList.get(key).split(separator)));
+            } else if (!oldClaimList.containsKey(key) && newClaimList.containsKey(key)) {
+                simpleMultiValuedClaimsToBeAdded.put(key, Arrays.asList(newClaimList.get(key).split(separator)));
+            }
+        }
+
+        // Prepare user claims expect multi-valued claims to be added, deleted and modified.
```
Comment (somindatommy) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java:3878]: Can we move this to the multi-line block?

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java [hunk 10]
```diff
@@ -3672,6 +3833,116 @@ private void updateUserClaims(User user, Map<String, String> oldClaimList,
...
+        }
+    }
+
+    /**
+     * Convert the claim values to list of strings. Main usage is for multi-valued claims.
+     *
+     * @param claimMap Map of claim URIs against claim value as string.
+     * @return Map of claim URIs against claim value as a list of string.
+     */
+    private Map<String, List<String>> convertClaimValuesToList(Map<String, String> claimMap) {
```
Comment (somindatommy) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java:3932]: `buildMapOfClaims` WDYT? In the method description we can say what we are trying to achieve.

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java [hunk 11]
```diff
@@ -3672,6 +3833,116 @@ private void updateUserClaims(User user, Map<String, String> oldClaimList,
...
+            carbonUM.setUserClaimValuesWithID(user.getId(), userClaimsToBeModified, null);
+        } else {
+            carbonUM.setUserClaimValuesWithID(user.getId(), convertClaimValuesToList(oldClaimList),
+                    simpleMultiValuedClaimsToBeAdded, simpleMultiValuedClaimsToBeRemoved,
+                    convertClaimValuesToList(userClaimsToBeModified), null);
+        }
+    }
+
+    /**
+     * Convert the claim values to list of strings. Main usage is for multi-valued claims.
```
Comment (somindatommy) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java:3927]: Shall we not mention about the usage of the method? WDYT

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java (hunk 5 above)
Comment (AnuradhaSK) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java:934]: There may be situations that come to this point other than the own person tries to change hi/her username. eg: If an admin tries to modify the username of another person, this " `User: <old username> attempting`" is misleading.
WDYT about this` BAD REQUEST. Attempting to modify username: <old username> as <new username>.`

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java (hunk 10 above)
Comment (AnuradhaSK) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java:3932]: I feel like `buildMapOfClaims` doesn't suite as the method signature even the description says the method behavior. The reason : the parameter which passes into this method also a claim map.

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java (hunk 11 above)
Comment (AnuradhaSK) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java]: addressed in https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/266/commits/d419b5d649f8bc9093b6c4143d0fa188cce5ad99

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java (hunk 9 above)
Comment (AnuradhaSK) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java]: addressed in https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/266/commits/d419b5d649f8bc9093b6c4143d0fa188cce5ad99

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java (hunk 8 above)
Comment (AnuradhaSK) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java]: addressed in https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/266/commits/d419b5d649f8bc9093b6c4143d0fa188cce5ad99

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java (hunk 7 above)
Comment (AnuradhaSK) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java]: addressed in https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/266/commits/d419b5d649f8bc9093b6c4143d0fa188cce5ad99

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java (hunk 6 above)
Comment (AnuradhaSK) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java]: addressed in https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/266/commits/d419b5d649f8bc9093b6c4143d0fa188cce5ad99

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java (hunk 4 above)
Comment (AnuradhaSK) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java]: addressed in https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/266/commits/d419b5d649f8bc9093b6c4143d0fa188cce5ad99

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java (hunk 3 above)
Comment (AnuradhaSK) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java]: addressed in https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/266/commits/d419b5d649f8bc9093b6c4143d0fa188cce5ad99

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java (hunk 2 above)
Comment (AnuradhaSK) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java]: addressed in https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/266/commits/d419b5d649f8bc9093b6c4143d0fa188cce5ad99

**Review by somindatommy — APPROVED** [review]

**Pull request merged**